### ISSUE TITLE
Compiling `petgraph` with `-D warnings` fails

### ISSUE THREAD

**Issue by grtlr** [petgraph/petgraph#499]
Compiling `petgraph` with `-D warnings` fails, which prevents depending projects from using `RUSTFLAGS="-D warnings -D missing-docs" cargo build` as part of their CI workflow, for example.

I think it has something to do with how `indexset` did semantic versioning.

### Steps To Reproduce

```
RUSTFLAGS="-D warnings -D missing_docs" cargo build
```

### Actual Results

```
error[E0107]: this struct takes 3 generic arguments but 2 generic arguments were supplied
  --> src/graphmap.rs:62:12
   |
62 |     nodes: IndexMap<N, Vec<(N, CompactDirection)>>,
   |            ^^^^^^^^ -  -------------------------- supplied 2 generic arguments
   |            |
   |            expected 3 generic arguments
   |
note: struct defined here, with 3 generic parameters: `K`, `V`, `S`
```

### Expected Results

No warnings / errors emitted.

**Comment by workingjubilee** [comment]
That is a type error. Such a compiler error will always fail. Using or not using `-Dwarnings` should have no effect on it. In addition, I can pull and build petgraph@3e1a0c48bda1392ef38adeca1e878b3dfb0ea04c, the last release commit. This is not an issue with petgraph.

**Comment by grtlr** [comment]
The following still produces these warnings:

```sh
git checkout 3dfb0ea04c
cargo build # compiles with `#[warn(dead_code)]` on field `first_error`
RUSTFLAGS="-D warnings -D missing_docs" cargo build # compiles with the error above
```

However, if I set the `std` feature on `indexmap` in my `Cargo.toml`, `petgraph` compiles with `-D warnings`.

I'm on macOS (Apple Silicon M1 Pro). If there is anything that I can do to help figure this out, just let me know!

**Comment by workingjubilee** [comment]
I noted whether I could compile the crate at all, because when dependencies are compiled with Rust, they are compiled with `--cap-lints=allow`. Thus, as I understand it, denying warnings or not should not affect dependencies. If it does, I am pretty sure that would be a bug in Cargo, or perhaps that you should reconsider setting `-Dwarnings` in RUSTFLAGS and thus as a directive for the entire build.

**Comment by XVilka** [comment]
Still happens. I ignored missing docs for now because there are many cases, let's focus on warnings first:
```
RUSTFLAGS="-D warnings" cargo build
warning: /Users/xvilka/Misc/petgraph/serialization-tests/Cargo.toml: no edition set: defaulting to the 2015 edition while the latest is 2021
   Compiling equivalent v1.0.1
   Compiling hashbrown v0.14.5
   Compiling fixedbitset v0.5.7
   Compiling indexmap v2.5.0
   Compiling petgraph v0.6.5 (/Users/xvilka/Misc/petgraph)
error: method `node_bound_with_dummy` is never used
   --> src/algo/matching.rs:108:8
    |
106 | trait WithDummy: NodeIndexable {
    |       --------- method in this trait
107 |     fn dummy_idx(&self) -> usize;
108 |     fn node_bound_with_dummy(&self) -> usize;
    |        ^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `-D dead-code` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(dead_code)]`

error: field `first_error` is never read
   --> src/csr.rs:134:5
    |
133 | pub struct EdgesNotSorted {
    |            -------------- field in this struct
134 |     first_error: (usize, usize),
    |     ^^^^^^^^^^^
    |
    = note: `EdgesNotSorted` has derived impls for the traits `Debug` and `Clone`, but these are intentionally ignored during dead code analysis

error: could not compile `petgraph` (lib) due to 2 previous errors
```

**Comment by starovoid** [comment]
Hi, I've fixed the `-D warnings` in PR #701.

**Comment by XVilka** [comment]
Fixed thanks to @starovoid!
